rustc fails to see that trait bound is satisfied

### Code

I tried this code:
```rust
use std::marker::PhantomData;

pub struct MyGenericType<T> {
    _marker: PhantomData<*const T>,
}

pub struct MyNonGenericType;

impl<T> From<MyGenericType<T>> for MyNonGenericType {
    fn from(_: MyGenericType<T>) -> Self {
        todo!()
    }
}

pub trait MyTrait {
    const MY_CONSTANT: i32;
}

impl<T> MyTrait for MyGenericType<T>
where
    Self: Into<MyNonGenericType>,
{
    const MY_CONSTANT: i32 = 1;
}

impl<T> MyGenericType<T> {
    const MY_OTHER_CONSTANT: i32 = <MyGenericType<T> as MyTrait>::MY_CONSTANT;
}

```

I expected to see this happen:
It should compile successfully.

Instead, this happened: 
The compiler doesn't realize that `Into<MyNonGenericType>` is implemented for all `MyGenericType<T>` and reports an error.

```
error[E0277]: the trait bound `MyNonGenericType: From<MyGenericType<T>>` is not satisfied
  --> src/lib.rs:27:36
   |
27 |     const MY_OTHER_CONSTANT: i32 = <MyGenericType<T> as MyTrait>::MY_CONSTANT;
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<MyGenericType<T>>` is not implemented for `MyNonGenericType`
   |
   = note: required because of the requirements on the impl of `Into<MyNonGenericType>` for `MyGenericType<T>`
note: required because of the requirements on the impl of `MyTrait` for `MyGenericType<T>`
  --> src/lib.rs:19:9
   |
19 | impl<T> MyTrait for MyGenericType<T>
   |         ^^^^^^^     ^^^^^^^^^^^^^^^^
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
26 | impl<T> MyGenericType<T> where MyNonGenericType: From<MyGenericType<T>> {
   |                          ++++++++++++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `regression` due to previous error
```

Interestingly enough, removing the `where Self: Into<MyNonGenericType>,` bound makes this compile again.

### cargo-bisect-rustc report
searched nightlies: from nightly-2021-12-01 to nightly-2021-12-25
regressed nightly: nightly-2021-12-14
searched commit range: https://github.com/rust-lang/rust/compare/6bda5b331cfe7e04e1fe348c58a928fc2b650f4f...8f117a77d0880ed59afcc1a19c72ec5c1e44b97c
regressed commit: https://github.com/rust-lang/rust/commit/22f8bde876f2fa9c5c4e95be1bce29cc271f2b51

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.1</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --test-dir=. --start=2021-12-01 --end=2021-12-25 
```
</details>

@rustbot modify labels: +regression-from-stable-to-nightly

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"weirane"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/src/test/ui/traits/issue-92292.rs
+++ b/src/test/ui/traits/issue-92292.rs
@@ -1,0 +1,32 @@
+// check-pass
+
+use std::marker::PhantomData;
+
+pub struct MyGenericType<T> {
+    _marker: PhantomData<*const T>,
+}
+
+pub struct MyNonGenericType;
+
+impl<T> From<MyGenericType<T>> for MyNonGenericType {
+    fn from(_: MyGenericType<T>) -> Self {
+        todo!()
+    }
+}
+
+pub trait MyTrait {
+    const MY_CONSTANT: i32;
+}
+
+impl<T> MyTrait for MyGenericType<T>
+where
+    Self: Into<MyNonGenericType>,
+{
+    const MY_CONSTANT: i32 = 1;
+}
+
+impl<T> MyGenericType<T> {
+    const MY_OTHER_CONSTANT: i32 = <MyGenericType<T> as MyTrait>::MY_CONSTANT;
+}
+
+fn main() {}

